### PR TITLE
Add axis labels

### PR DESCRIPTION
**Pull request by smasahir**
Axis labels are shown

### CODE REVIEW

File: src/mixins/controls.js [hunk 1]
```diff
@@ -1,5 +1,9 @@
 import * as THREE from "three";
+import {CSS2DRenderer} from "three-css2drender"
```
Comment (timurbazhirov) [inline src/mixins/controls.js]: Let's try and import CSS2DRenderer from three.js directly, and update the three.js version, if necessary

File: src/mixins/controls.js [hunk 2]
```diff
@@ -1,5 +1,9 @@
...
 import {UtilsMixin} from "./utils";
+import {addAxisLabels} from "./labels";
+import { materialsToThreeDSceneData } from "../utils";
```
Comment (timurbazhirov) [inline src/mixins/controls.js]: We should establish the coding style according to https://github.com/Exabyte-io/webstorm-settings

File: src/mixins/controls.js [hunk 3]
```diff
@@ -136,17 +140,38 @@ const OrbitControlsMixin = (superclass) => class extends superclass {
         this.render();
     }
 
+
```
Comment (timurbazhirov) [inline src/mixins/controls.js]: Remove

File: src/mixins/controls.js [hunk 4]
```diff
@@ -136,17 +140,38 @@ const OrbitControlsMixin = (superclass) => class extends superclass {
...
      * Initialize a "picture-in-picture" Axes Helper to visualize camera movements around the object.
      */
     initSecondAxes() {
         const length = 100;
         const containerDimension = 100;
-
         this.renderer2 = this.getWebGLRenderer({alpha: true});
         this.renderer2.setClearColor("#FFFFFF", 0);
         this.renderer2.setSize(containerDimension, containerDimension);
+		this.renderer2.domElement.style.position = 'absolute';
```
Comment (timurbazhirov) [inline src/mixins/controls.js]: Unnecessary tab - see note above

File: src/mixins/controls.js [hunk 5]
```diff
@@ -136,17 +140,38 @@ const OrbitControlsMixin = (superclass) => class extends superclass {
...
+		this.renderer2CSS2D = new CSS2DRenderer();
+        this.renderer2CSS2D.setSize(containerDimension, containerDimension);
+		this.renderer2CSS2D.domElement.style.position = 'absolute';
+		this.renderer2CSS2D.domElement.style.right = '10px';
+		this.renderer2CSS2D.domElement.style.pointerEvents = 'none';
+        this.container.prepend(this.renderer2CSS2D.domElement);
+
+        //axis label X
+        var coordX = {x:length * 1.2, y:0, z:0}
+        const labelX = addAxisLabels('X', coordX, 'rgb(255,255,255)', 'rgb(0,0,0)');
```
Comment (timurbazhirov) [inline src/mixins/controls.js]: Let's put labelColor to settings.js file

File: src/mixins/controls.js [hunk 6]
```diff
@@ -160,9 +185,25 @@ const OrbitControlsMixin = (superclass) => class extends superclass {
...
+        /*
         this.scene2.x = x;
         this.scene2.y = y;
         this.scene2.z = z;
+
+        //axis labels
+        this.scene2.labelX = labelX;
+        this.scene2.labelY = labelY;
+        this.scene2.labelZ = labelZ;
+        */
```
Comment (timurbazhirov) [inline src/mixins/controls.js]: Let's clean up the above

File: src/mixins/controls.js [hunk 7]
```diff
@@ -174,13 +215,16 @@ const OrbitControlsMixin = (superclass) => class extends superclass {
     }
 
     showSecondAxes() {
-        const secondAxes = [this.scene2.x, this.scene2.y, this.scene2.z].filter(x => x);  // assert no `undefined`;
-        this.scene2.add(...secondAxes);
+        //const secondAxes = [this.scene2.x, this.scene2.y, this.scene2.z, this.scene2.labelX, this.scene2.labelY, this.scene2.labelZ ].filter(x => x);  // assert no `undefined`;
```
Comment (timurbazhirov) [inline src/mixins/controls.js]: Let's move the code from `initSecondAxes` here instead

File: src/mixins/labels.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    const text = document.createElement('div');
+    text.className ='label';
+    text.style.color = color;
+    text.style.backgroundColor = background;
+    text.textContent = symbol;
+    const label = new CSS2DObject( text );
+    label.position.copy(coord);
+
+    return label;
+}
```
Comment (timurbazhirov) [inline src/mixins/labels.js]: No newline

**Comment by timurbazhirov** [comment]
@masahirosaito - the commit messages above are duplicated ("Axis labels are shown" is repeated several times). Please use some more meaningful messages in the future, like "Updated styles" or "Respond to code review comments"

File: package.json [hunk 9]
```diff
@@ -40,7 +40,7 @@
         "sprintf-js": "^1.1.2",
         "static-kdtree": "^1.0.2",
         "sweetalert": "^2.1.2",
-        "three": "^0.90.0",
+        "three": "^0.111.0",
```
Comment (timurbazhirov) [inline package.json]: Have you tried [running tests](https://github.com/Exabyte-io/wave.js#tests) with the new Three.js version and it was successful?

File: src/mixins/controls.js [hunk 10]
```diff
@@ -173,14 +200,12 @@ const OrbitControlsMixin = (superclass) => class extends superclass {
...
 
-    showSecondAxes() {
-        const secondAxes = [this.scene2.x, this.scene2.y, this.scene2.z].filter(x => x);  // assert no `undefined`;
-        this.scene2.add(...secondAxes);
-    }
 
     hideSecondAxes() {
-        const secondAxes = [this.scene2.x, this.scene2.y, this.scene2.z].filter(x => x);  // assert no `undefined`;
-        this.scene2.remove(...secondAxes);
+        this.scene2.remove(this.scene2.secondAxesGroup);
```
Comment (timurbazhirov) [inline src/mixins/controls.js]: We should keep `showSecondAxes` function, and move some of the code from `enableOrbitControls` into it

File: src/mixins/controls.js [hunk 11]
```diff
@@ -146,7 +148,26 @@ const OrbitControlsMixin = (superclass) => class extends superclass {
         this.renderer2 = this.getWebGLRenderer({alpha: true});
         this.renderer2.setClearColor("#FFFFFF", 0);
         this.renderer2.setSize(containerDimension, containerDimension);
+        this.renderer2.domElement.style.position = 'absolute';
+        this.renderer2.domElement.style.right = '10px';
```
Comment (timurbazhirov) [inline src/mixins/controls.js]: Let's move lines 148-152 to a separate function: `initRenderer2`

File: src/mixins/labels.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import {CSS2DObject} from "three/examples/jsm/renderers/CSS2DRenderer.js"
+
+export const addAxisLabels = function(symbol, coord, color, background) {
+    const text = document.createElement('div');
```
Comment (timurbazhirov) [inline src/mixins/labels.js]: Docstring?

File: src/mixins/controls.js [hunk 13]
```diff
@@ -146,7 +148,26 @@ const OrbitControlsMixin = (superclass) => class extends superclass {
...
         this.container.prepend(this.renderer2.domElement);
+        //For axis labels
+        this.renderer2CSS2D = new CSS2DRenderer();
+        this.renderer2CSS2D.setSize(containerDimension, containerDimension);
+        this.renderer2CSS2D.domElement.style.position = 'absolute';
+        this.renderer2CSS2D.domElement.style.right = '10px';
+        this.renderer2CSS2D.domElement.style.pointerEvents = 'none';
+        this.container.prepend(this.renderer2CSS2D.domElement);
```
Comment (timurbazhirov) [inline src/mixins/controls.js]: Lines 155-160 to a separate function `initRendered2CSS2D`

File: src/mixins/controls.js [hunk 14]
```diff
@@ -160,9 +181,15 @@ const OrbitControlsMixin = (superclass) => class extends superclass {
         this.camera2 = new THREE.PerspectiveCamera(50, 1, 1, 1000);
         this.camera2.up = this.camera.up;
         // saving axes helpers inside the scene object itself for further re-use in `hide*` method
-        this.scene2.x = x;
-        this.scene2.y = y;
-        this.scene2.z = z;
+        const secondAxesGroup = new THREE.Group();
+        secondAxesGroup.add(x);
```
Comment (timurbazhirov) [inline src/mixins/controls.js]: `secondAxesGroup.add(...[x, y, z, labelX, labelY, labelZ])`

File: src/mixins/controls.js [hunk 15]
```diff
@@ -160,9 +181,15 @@ const OrbitControlsMixin = (superclass) => class extends superclass {
...
+        secondAxesGroup.add(y);
+        secondAxesGroup.add(z);
+        secondAxesGroup.add(labelX);
+        secondAxesGroup.add(labelY);
+        secondAxesGroup.add(labelZ);
+        this.scene2.secondAxesGroup = secondAxesGroup;
+        console.log(this.scene2);
```
Comment (timurbazhirov) [inline src/mixins/controls.js]: Remove console.log

File: src/mixins/controls.js [hunk 16]
```diff
@@ -146,7 +148,26 @@ const OrbitControlsMixin = (superclass) => class extends superclass {
...
+        //For axis labels
+        this.renderer2CSS2D = new CSS2DRenderer();
+        this.renderer2CSS2D.setSize(containerDimension, containerDimension);
+        this.renderer2CSS2D.domElement.style.position = 'absolute';
+        this.renderer2CSS2D.domElement.style.right = '10px';
+        this.renderer2CSS2D.domElement.style.pointerEvents = 'none';
+        this.container.prepend(this.renderer2CSS2D.domElement);
+
+        //axis label
+        const coordX = { x: length * 1.2, y: 0, z: 0 }
```
Comment (timurbazhirov) [inline src/mixins/controls.js]: No shortened names, please `coord` -> `labelCoordinateX`

File: src/mixins/controls.js [hunk 17]
```diff
@@ -173,14 +200,12 @@ const OrbitControlsMixin = (superclass) => class extends superclass {
...
-        this.scene2.add(...secondAxes);
-    }
 
     hideSecondAxes() {
-        const secondAxes = [this.scene2.x, this.scene2.y, this.scene2.z].filter(x => x);  // assert no `undefined`;
-        this.scene2.remove(...secondAxes);
+        this.scene2.remove(this.scene2.secondAxesGroup);
+        while (this.renderer2CSS2D.domElement.firstChild) {
+            this.renderer2CSS2D.domElement.removeChild(this.renderer2CSS2D.domElement.firstChild);
+        }
```
Comment (timurbazhirov) [inline src/mixins/controls.js]: I don't understand why we need the while loop also - this is a potential performance concern. Let's just organize all objects in groups and remove groups directly

File: src/settings.js [hunk 18]
```diff
@@ -27,6 +27,7 @@ export default {
     // general
     backgroundColor: "#202020",
     defaultColor: "#CCCCCC",
+    labelColor: "rgb(255,255,255)",
```
Comment (timurbazhirov) [inline src/settings.js]: Other colors are defined in HEX - can we use it instead or `rgb`??

File: src/wave.js [hunk 19]
```diff
@@ -209,6 +209,8 @@ export class Wave extends mix(WaveBase).with(
     render() {
         this.renderer.render(this.scene, this.camera);
         this.renderer2 && this.renderer2.render(this.scene2, this.camera2);
+        //For axis labels
```
Comment (timurbazhirov) [inline src/wave.js]: There should always be a space after a comment starter `//For` -> `// For`

File: package.json [hunk 20]
```diff
@@ -40,7 +40,7 @@
...
         "three-orbit-controls": "^82.1.0",
         "three-transform-controls": "^1.0.4",
```
Comment (timurbazhirov) [inline package.json]: We should also make sure that the orbit controls and transform controls are compatible with the new THREE.js version - see the notion at the end of https://threejs.org/docs/index.html#manual/en/introduction/Import-via-modules

File: package.json (hunk 9 above)
Comment (masahirosaito) [reply package.json]: Running test of master branch went well with  the new Three.js version.
But, that  of add_axis_labels branch is  not successful.

I am looking for the cause of the following error:
  ● Test suite failed to run

    /home/msaito/GitHub/wave.js/node_modules/three/examples/jsm/renderers/CSS2DRenderer.js:5
    import {
    ^^^^^^
    SyntaxError: Cannot use import statement outside a module

Comment (masahirosaito) [reply package.json]: JSET seems to be unable to import  ES6 modules.
Although I  tried to modified .babelrc to be able to import ES6 modules, it didn't improve.


If the "three-css2drender" package is used as in the beginning https://github.com/Exabyte-io/wave.js/pull/38, the test is successful.

**Pull request closed without merging**